Retrieve only a summary of the condor jobs in DrainStatusPoller

amaltaro opened the pull request:
Fixes #8779 

#### Status
ready

#### Description
Query condor schedd only once, and ask for a summary of the condor jobs (instead of two queries, for idle and running jobs). It should also drastically decrease the memory footprint, since it no longer returns all the job classads.

#### Is it backward compatible (if not, which system it affects?)
yes

#### Related PRs
none

#### External dependencies / deployment changes
none

amaltaro committed:
Retrieve only a summary of the condor jobs in DrainStatusPoller

Log the query args; fetch absolute number of jobs

Remove PyCondorAPI import from DrainStatusAPI; fetch it from DrainStatusPoller

Put condor_status placeholder in the DrainStatusAPI

Move condor summary back to DrainStatusAPI; fetch total jobs in Drain poller

remove updateCondorDrainStats call

amaltaro committed:
fix emulator signature

Mock getCondorJobsSummary method

amaltaro commented:
I'm testing this patch on submit3, so far so good. Thread execution time went down from ~55secs to just under a second as well.
I also have the impression that this might fix the issue: https://github.com/dmwm/WMCore/issues/8779 , since we no longer retrieve and load all the job classsads in memory.

todor-ivanov commented on src/python/WMComponent/AgentStatusWatcher/DrainStatusAPI.py:
```diff
@@ -46,20 +46,26 @@ def checkWorkflows(self):
+            elif jobs is None:
+                results[state.lower()] = "unknown (schedd query error)"
+            elif totalOnly:
+                results += int(jobs[0].get(state))
```
And actually this should be exactly the way we should go. Because this will also be a softer query to the schedd, since it does not try to return the full list of classads. Here could be one example [1] how this is usually used in  the command line options and in the python bindings respectively. And calling the current function with `TotalOnly=True` is again a full new unique query, which in no way benefits from the fact that you unify the list  of states we query for in a single call.  So we gain nothing by keeping it under the same function, instead of confusion. 

[1]
https://htcondor.readthedocs.io/en/latest/man-pages/condor_q.html
Looking at option `-total`.  And what we should try with python bindings is giving that option as a parameter to the `htcondor.schedd.query()` method. like:

https://htcondor.readthedocs.io/en/latest/apis/python-bindings/api/htcondor.html#htcondor.Schedd.query
https://htcondor.readthedocs.io/en/latest/apis/python-bindings/api/htcondor.html#htcondor.QueryOpts.SummaryOnly

todor-ivanov replied:
Since we are on the topic, I believe the large memory footprint is due to the fact that in the `condorAPI` we use `htcondor.Schedd.query()` instead of `htcondor.schedd.xquery()`. It is mentioned in the warning in the description of the `query()` method in the HTCondor documentation cited above.

todor-ivanov commented on src/python/WMComponent/AgentStatusWatcher/DrainStatusPoller.py:
```diff
@@ -174,8 +174,8 @@ def checkSpeedDrainThresholds(self):
-        # get condor jobs
-        jobs = self.condorAPI.getCondorJobs("", [])
+        # get a summary of the condor jobs
+        jobs = self.drainAPI.checkCondorStates(totalOnly=True)
```
If my previous comment is to be taken into account, this call here should simply be converted to a call of a separate method, instead to a call of the same method with additional flag.

todor-ivanov commented on src/python/WMCore/Services/PyCondor/PyCondorAPI.py:
```diff
@@ -22,22 +22,31 @@ def __init__(self):
         self.coll = htcondor.Collector()
 
-    def getCondorJobs(self, constraint, attr_list):
+    def getCondorJobs(self, constraint='true', attrList=None, opts="SummaryOnly"):
```
looking at this `opts="SummaryOnly"`, I realize you were already aware of the option path I was previously talking about. And it is quite good you added it here, so I would suggest to use it in order to implement my suggestion above. Thanks for that one -  I really like it!

todor-ivanov reviewed (changes requested):
Thanks @amaltaro for this PR ! These suggestions of yours are for sure improving the things a lot. I have put few comments inline though, which I think  you should go through and maybe consider my suggestions.

amaltaro commented:
Repurposing this PR to actually fixing https://github.com/dmwm/WMCore/issues/8779 instead of #10170, since I fail to fix the latter.

amaltaro commented:
@todor-ivanov Todor, after your review, I have pushed two other commits that should be addressing your concerns. I still have to squash them and run a final test, but it would be great to have your feedback before I do so. Thanks

todor-ivanov commented on src/python/WMComponent/AgentStatusWatcher/DrainStatusPoller.py:
```diff
@@ -170,27 +170,44 @@ def checkSpeedDrainThresholds(self):
+
+        results = 0
+        if jobs:
+            for state in self.condorStates:
```
Do we care only about `Idle` and `Running` jobs here? What about the rest of the states? If we are about to drain a schedd my understanding is that we should suppose that it is free of not only `Idle` && `Running` jobs but also the ones in `Hold`. And there could be many reasons for a job to be in `Hold`, some of which may even require some debugging.

amaltaro replied:
Yes, here we should consider only Idle/Running. Because depending on their total number, we bump their priority or not. So we do not want to affect this decision by jobs lingering around in the `Held` (5) or `Removed` (3) states.
It might be useful though in the drain stats... let me check.

todor-ivanov commented on src/python/WMCore/Services/PyCondor/PyCondorAPI.py:
```diff
@@ -18,34 +17,64 @@
+        msg = "Querying condor schedd with params: constraint=%s, attrList=%s, limit=%s, opts=%s"
+        logging.info(msg, constraint, attrList, limit, opts)
+        try:
+            return self.schedd.xquery(constraint, attrList, limit, opts=opts)
```
This is great!

todor-ivanov commented on src/python/WMCore/Services/PyCondor/PyCondorAPI.py:
```diff
@@ -18,34 +17,64 @@
+            self.recreateSchedd()
+
+        # if we hit another exception, let it be raised up in the chain
+        return self.schedd.xquery(constraint, attrList, limit, opts=opts)
```
This one too!

todor-ivanov commented on src/python/WMCore/Services/PyCondor/PyCondorAPI.py:
```diff
@@ -18,34 +17,64 @@
+        Retrieves a job summary from the HTCondor Schedd object
+        :return: a list of classads representing the matching jobs, or None if failed
         """
+        ### NOTE: SummaryOnly does not work with xquery method
```
I know in this case it doesn't matter if the background query is `query` or `xquery` because we do expect only the total numbers to be provided  by HTcondor, but just for my curiosity, is it documented in the HTCondor the fact that this option is not supported with `xquery` or you had to discover it yourself?

amaltaro replied:
I haven't seen any documentation on that, so I had to find it the hard way.
If it isn't too expensive to file an HTCondor issue, I might do so (even though we are not using their latest version...)

todor-ivanov reviewed (approved):
Thanks @amaltaro! You did address the changes requested. There are a comment or two I've put inline, but those are just clarifying questions. The code look perfectly good as it is.

amaltaro commented:
Thanks for your review and comments, Todor.
Looking again at the output of the SummaryOnly option:
```
[[ SchedulerHeld = 0; SchedulerRemoved = 0; SchedulerRunning = 0; SchedulerIdle = 0; ServerTime = 1619512078; Held = 0; AllusersSchedulerRemoved = 0; AllusersSchedulerCompleted = 0; ErrorCode = 0; SchedulerCompleted = 0; AllusersSchedulerJobs = 0; AllusersRemoved = 0; SchedulerJobs = 0; Running = 2716; AllusersHeld = 0; AllusersSchedulerHeld = 0; Removed = 0; AllusersCompleted = 0; Suspended = 0; AllusersJobs = 4690; AllusersRunning = 2716; MyType = "Summary"; AllusersSchedulerIdle = 0; Jobs = 4690; AllusersIdle = 1974; Idle = 1974; AllusersSuspended = 0; AllusersSchedulerRunning = 0; Completed = 0 ]]
```

we could have a minor improvement to the drain stats, by also providing the total number of `Jobs` in the queue. But then it requires changes to the couchapps as well. I'd rather leave that for another moment...

amaltaro commented:
I did some quick tests and things look fine. I have also decided not to backport it to the current production release, since it does not fix the draining issue. Thanks for the review Todor.
